### DISCUSSION TITLE
requirements

### DISCUSSION THREAD

**Discussion by otzen** [nukebloodaxe/Ironseed-py#9] [Q&A]
I just tried downloading an following the instructions in the readme, but it fails at the pip install, as the requirements list a lot of system libraries like io and time. And pip fails because it cannot find such on pypi.
Does this work at your end?
My updated list that will pass looks like this
###### Requirements without Version Specifiers ######
pygame
#io
#os
#math
#random
#sys
#time
numpy
#

**Comment by nukebloodaxe** [comment]
Well, these are the requirements for this to run.
io is used for various input/output operations regarding file access.
os is used for handling pathname agnostic access.
math is used for internal math routines.
random is used for random number generation; as well as random seed support for roguelike game generation.
sys is used for system routines.
time is used by a variety of system timers, including the ironseed stopwatch.
numpy is not actually used... yet.

So, I have tested running on a variety of Linux systems (OpenSUSE, Tumbleweed, Ubuntu, Debian), Windows, all x86, and also Oracles AM1 systems under Ubuntu.

In order to diagnose why your system does not support these libraries, I will need more information. Preferably the OS, python version, and the hardware running it; if you're trying to run this in Android, I honestly don't think that will work, I've not targeted it.

For a quicker, and easier time, I recommend installing Anaconda, and running this via spyder; almost all the needed libraries are working out of the box, and you only need to do a pip install of pygame in the spyder terminal.  [https://www.anaconda.com/products/individual](https://www.anaconda.com/products/individual)

Alternatively, I highly recommend installing the required python libraries using your OS package manager, it's total chaos out there otherwise in terms of libraries; I confirmed that method works fine in OpenSUSE, Tunbleweed, Ubuntu and Debian.

Note:  I find it really odd pip did not find those libraries, they are pretty basic things required for anything beyond pure language functions, especially math, random and time.  I also did a quick check to see if it all runs still, which fired up and got going quite well in Spyder running in a current TumbleWeed build.

**Reply by otzen** [reply]
Maybe my comment was not clear, Yes your code code uses the io, random, math ..... libraries.
The problem is the requirements file and the use of pip. specifically this command from your readme.md file
`python3 -m pip install -r requirements.txt`
This will fail. ANd the reason it will fail is that pip and pypi is serving thirdparty libraries for python, like pygame. and libraries like io is not third party, they are part of the standard library, ( [as you can find it here](https://docs.python.org/3/library/io.html) ).
Thus when I install python 3 i also get those libraries, so I do not need pip to fetch them, and when i do ask pip to fetch it, it gets confused as it is only looking thirdparty libs on pypi.org.
That is also why you should not list those libs in the requirements file, because this is only ment for ecternal requirements, as soon as you say python3 you implicit provide all the standard libs like io, random, time a.s.o.
So after having installed python itself you only need to install pygame and numpy, this is what requierements.txt and pip is there for.

**Reply by nukebloodaxe** [reply]
Implicit is not explicit.  I have commented out the lines as a concession to those who use pip as an automatic mechanism, but the requirements will remain listed in the file so that people working with other architectures are aware of what is actually required.  One should not assume the default libraries are actually installed defaults on all platforms; things can change in the future, and someone may also attempt to run this under python 2.7, which won't work, and it's preferable for them to know why without analysing the code to see it is Python 3 dependent.

**Comment by otzen** [comment]
That's fine, Also as you actually give the recipe on how to use it with pip in your readme.md file.

But let me rule out some misunderstandings here.

First: It is hard for me to predict how this is being read. So if you get the feeling I am talking down to you, please don't. That is not my intention. You write somewhere that this project is a learning task for you, a splendid way to learn if i may say so. I just try to assist a little in that learning journey :-).

Second: The system libraries is not a random set of libraries that varies from platform to platform. it is specified by the Python organization. So if you install python3 you **_always_** get those libraries included, no matter if you do it on windows, Linux, BSD or what ever platform.

Regarding Python3 vs. Python2.7, the differences here is way beyond the included system libraries, there is changes to the language itself, some commands has arrived in Py3, others has gone or changed interface.
The most obvious one is the print command in Py2 this was allowed
`print 'Hello'`
This is no more legal in Py3 here you have to write
`print ('Hello')`
Example of a new thing in p3 is the f-strings. That is a new and way smarter way to format strings.
Both of the above is not solved by just including a library.

And if you really want to list all the dependencies including the ones to system libs, well how about the dependencies of those libs, for example `os` uses the `abc` library a.s.o.
To avoid making the requirements file a huge list, it is standard to just mention the top libraries, and then let pip take care of the dependencies from them, that is actually the essence of a package managers job.
And by saying Python3 in your readme.md You have pointed out the top lib of those standard libraries.

Note that pip is not involved in installing python3 it-self, as it it-self is a python program. only third party libraries.
